wordpress-mobile/WordPress-iOS
# Site Settings in SwiftUI: Privacy

**HACK week:** update Site Settings to use SwiftUI

As I was working on the [original PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/20874), I realized two things:

- `SiteSettingsViewController.m` has a few extensions and much more code than I initially estimated. It's not feasible to rewrite the whole thing within a week.
- Most of the cells in `SiteSettingsViewController` show custom UIKit-based screen. You can't really use any of the native SwiftUI form components. There is too much UIKit-SwiftUI dance needed to rewrite this screen without rewriting the leaf screens first.

So I decided to salvage what's left from the [original PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/20874) and pivot to replacing the "leaf" screens and adding all the reusable components and other infra for SwiftUI.

Changes:
- Add `SettingsPicker` to replace `SettingsSelectionViewController` (`SwiftUI.Picker` has a non-customizable design that can't be changed to match the WP design)
- Update Site Settings to use the new picker
- Rewrite SiteVisibilityHelpers in Swift

## To test:

- Open Dashboard / Menu / Site Settings / Privacy
- Select a new value
- Verify that the settings is updated (check on the web or on another phone)
- Verify that hint (section footer) is updated
- Return to Site Settings
- Verify that Privacy value is updated

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/101070cd-d616-4a98-9fac-961e43f544b1

## Regression Notes
1. Potential unintended areas of impact: Site Settings / Privacy
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual test and updated existing unit tests
3. What automated tests I added (or what prevented me from doing so): test eligible privacy settings

PR submission checklist:


UI Changes testing checklist:
